File: 백준/Gold/2668. 숫자고르기/숫자고르기.swift
import Foundation

let N = Int(readLine()!)!
var numDict: [Int:Int] = [:]
for i in 1...N {
    numDict[i] = Int(readLine()!)!
}

var visited: [Bool] = Array(repeating: false, count: N+1)
var results: [Int] = []

for i in 1...N {
    if !results.contains(i) {
        var visitList = [i]
        dfs(node: i, visitList: &visitList)
        visited = Array(repeating: false, count: N+1)
    }
}

print(results.count)
results.sort()
results.forEach { print($0) }

func dfs(node: Int,  visitList: inout [Int]) {
    visited[node] = true
    let nextNode = numDict[node]!
    if visited[nextNode] { // 이미 방문
        if visitList.first == nextNode {
            results.append(contentsOf: visitList)
        }
    } else {
        visitList.append(nextNode)
        dfs(node: nextNode, visitList: &visitList)
    }
}

File: 백준/Gold/2668. 숫자고르기/숫자고르기.py
import sys
input = sys.stdin.readline

N = int(input())
numList = {i: int(input()) for i in range(1, N+1)}
results = []
visited = [False]*(N+1)

def dfs(node, visitList):
    visited[node] = True
    nextNode = numList[node]
    if visited[nextNode]: #다음 노드를이미 방문 했으면
        if visitList[0] == nextNode:
            results.extend(visitList)
    else:
        dfs(nextNode, visitList+[nextNode])
    
for i in range(1, N+1):
    if i not in results:
        dfs(i, [i])
        visited = [False]*(N+1)
print(len(results))
results.sort()
print(*results, sep="\n")